File: paddle/fluid/framework/executor.cc
/* Copyright (c) 2016 PaddlePaddle Authors. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License. */

#include "paddle/fluid/framework/executor.h"

#include <set>

#include "gflags/gflags.h"
#include "paddle/fluid/framework/channel.h"
#include "paddle/fluid/framework/feed_fetch_method.h"
#include "paddle/fluid/framework/feed_fetch_type.h"
#include "paddle/fluid/framework/lod_rank_table.h"
#include "paddle/fluid/framework/lod_tensor_array.h"
#include "paddle/fluid/framework/op_registry.h"
#include "paddle/fluid/framework/reader.h"
#include "paddle/fluid/platform/place.h"
#include "paddle/fluid/platform/profiler.h"

DECLARE_bool(benchmark);
DEFINE_bool(check_nan_inf, false,
            "Checking whether operator produce NAN/INF or not. It will be "
            "extremely slow so please use this flag wisely.");

namespace paddle {
namespace framework {
namespace {

std::mutex reduce_mu;
std::vector<std::string> runned;

// block id starts from 0. This id is used to represent the codeblock
// wrapping the first block 0.
int kProgramId = -1;
}  // namespace

struct ExecutorPrepareContext {
  ExecutorPrepareContext(const framework::ProgramDesc& prog, size_t block_id)
      : prog_(prog), block_id_(block_id) {}

  const framework::ProgramDesc& prog_;
  size_t block_id_;
  std::vector<std::unique_ptr<OperatorBase>> ops_;
};


Executor::Executor(const platform::Place& place) : place_(place) {}

static void CreateTensor(Variable* var, proto::VarType::Type var_type) {
  if (var_type == proto::VarType::LOD_TENSOR) {
    var->GetMutable<LoDTensor>();
  } else if (var_type == proto::VarType::SELECTED_ROWS) {
    var->GetMutable<SelectedRows>();
  } else if (var_type == proto::VarType::FEED_MINIBATCH) {
    var->GetMutable<FeedFetchList>();
  } else if (var_type == proto::VarType::FETCH_LIST) {
    var->GetMutable<FeedFetchList>();
  } else if (var_type == proto::VarType::STEP_SCOPES) {
    var->GetMutable<std::vector<framework::Scope>>();
  } else if (var_type == proto::VarType::LOD_RANK_TABLE) {
    var->GetMutable<LoDRankTable>();
  } else if (var_type == proto::VarType::LOD_TENSOR_ARRAY) {
    var->GetMutable<LoDTensorArray>();
  } else if (var_type == proto::VarType::PLACE_LIST) {
    var->GetMutable<platform::PlaceList>();
  } else if (var_type == proto::VarType::READER) {
    var->GetMutable<ReaderHolder>();
  } else if (var_type == proto::VarType::CHANNEL) {
    var->GetMutable<ChannelHolder>();
  } else if (var_type == proto::VarType::RAW) {
    // GetMutable will be called in operator
  } else {
    PADDLE_THROW(
        "Variable type %d is not in "
        "[LOD_TENSOR, SELECTED_ROWS, FEED_MINIBATCH, FETCH_LIST, "
        "LOD_RANK_TABLE, PLACE_LIST, READER, CHANNEL, RAW]",
        var_type);
  }
}

static void CheckTensorNANOrInf(const std::string& name,
                                const framework::Tensor& tensor) {
  if (tensor.memory_size() == 0) {
    return;
  }
  if (tensor.type().hash_code() != typeid(float).hash_code() &&
      tensor.type().hash_code() != typeid(double).hash_code()) {
    return;
  }
  PADDLE_ENFORCE(!framework::TensorContainsInf(tensor),
                 "Tensor %s contains Inf", name);
  PADDLE_ENFORCE(!framework::TensorContainsNAN(tensor),
                 "Tensor %s contains NAN", name);
}

void WaitOnPlace(const platform::Place& place) {
  platform::DeviceContextPool &pool = platform::DeviceContextPool::Instance();
  auto &dev_ctx = *pool.Get(place);
  dev_ctx.Wait();
}

void Executor::Run(const ProgramDesc& pdesc, Scope* scope, int block_id,
                   bool create_local_scope, bool create_vars) {
  platform::RecordBlock b(block_id);
  auto* ctx = Prepare(pdesc, block_id);
  RunPreparedContext(ctx, scope, create_local_scope, create_vars);
  delete ctx;
}

// Check whether the block already has feed operators and feed_holder.
// Return false if the block does not have any feed operators.
// If some feed operators have been prepended to the block, check that
// the info contained in these feed operators matches the feed_targets
// and feed_holder_name. Raise exception when any mismatch is found.
// Return true if the block has feed operators and holder of matching info.
static bool has_feed_operators(
    BlockDesc* block, std::map<std::string, const LoDTensor*>& feed_targets,
    const std::string& feed_holder_name) {
  size_t feed_count = 0;
  for (auto* op : block->AllOps()) {
    if (op->Type() == kFeedOpType) {
      feed_count++;
      PADDLE_ENFORCE_EQ(op->Input("X")[0], feed_holder_name,
                        "Input to feed op should be '%s'", feed_holder_name);
      std::string feed_target_name = op->Output("Out")[0];
      PADDLE_ENFORCE(
          feed_targets.find(feed_target_name) != feed_targets.end(),
          "Feed operator output name '%s' cannot be found in 'feed_targets'",
          feed_target_name);
    }
  }

  if (feed_count > 0) {
    PADDLE_ENFORCE_EQ(
        feed_count, feed_targets.size(),
        "The number of feed operators should match 'feed_targets'");

    // When feed operator are present, so should be feed_holder
    auto var = block->FindVar(feed_holder_name);
    PADDLE_ENFORCE_NOT_NULL(var, "Block should already have a '%s' variable",
                            feed_holder_name);
    PADDLE_ENFORCE_EQ(var->GetType(), proto::VarType::FEED_MINIBATCH,
                      "'%s' variable should be 'FEED_MINIBATCH' type",
                      feed_holder_name);
  }

  return feed_count > 0;
}

// Check whether the block already has fetch operators and fetch_holder.
// Return false if the block does not have any fetch operators.
// If some fetch operators have been appended to the block, check that
// the info contained in these fetch operators matches the fetch_targets
// and fetch_holder_name. Raise exception when any mismatch is found.
// Return true if the block has fetch operators and holder of matching info.
static bool has_fetch_operators(
    BlockDesc* block, std::map<std::string, LoDTensor*>& fetch_targets,
    const std::string& fetch_holder_name) {
  size_t fetch_count = 0;
  for (auto* op : block->AllOps()) {
    if (op->Type() == kFetchOpType) {
      fetch_count++;
      PADDLE_ENFORCE_EQ(op->Output("Out")[0], fetch_holder_name,
                        "Output of fetch op should be '%s'", fetch_holder_name);
      std::string fetch_target_name = op->Input("X")[0];
      PADDLE_ENFORCE(
          fetch_targets.find(fetch_target_name) != fetch_targets.end(),
          "Fetch operator input name '%s' cannot be found in 'fetch_targets'",
          fetch_target_name);
    }
  }

  if (fetch_count > 0) {
    PADDLE_ENFORCE_EQ(
        fetch_count, fetch_targets.size(),
        "The number of fetch operators should match 'fetch_targets'");

    // When fetch operator are present, so should be fetch_holder
    auto var = block->FindVar(fetch_holder_name);
    PADDLE_ENFORCE_NOT_NULL(var, "Block should already have a '%s' variable",
                            fetch_holder_name);
    PADDLE_ENFORCE_EQ(var->GetType(), proto::VarType::FETCH_LIST,
                      "'%s' variable should be 'FETCH_LIST' type",
                      fetch_holder_name);
  }

  return fetch_count > 0;
}

void Executor::Run(const ProgramDesc& program, Scope* scope,
                   std::map<std::string, const LoDTensor*>& feed_targets,
                   std::map<std::string, LoDTensor*>& fetch_targets,
                   const std::string& feed_holder_name,
                   const std::string& fetch_holder_name) {
  platform::RecordBlock b(kProgramId);
  auto* copy_program = new ProgramDesc(program);
  auto* global_block = copy_program->MutableBlock(0);

  if (!has_feed_operators(global_block, feed_targets, feed_holder_name)) {
    // create feed_holder variable
    auto* feed_holder = global_block->Var(feed_holder_name);
    feed_holder->SetType(proto::VarType::FEED_MINIBATCH);
    feed_holder->SetPersistable(true);

    // for (auto block : program.Block())

    int i = 0;
    for (auto& feed_target : feed_targets) {
      std::string var_name = feed_target.first;
      VLOG(3) << "feed target's name: " << var_name;

      // prepend feed op
      auto* op = global_block->PrependOp();
      op->SetType(kFeedOpType);
      op->SetInput("X", {feed_holder_name});
      op->SetOutput("Out", {var_name});
      op->SetAttr("col", {static_cast<int>(i)});
      op->CheckAttrs();

      i++;
    }
  }

  // map the data of feed_targets to feed_holder
  for (auto* op : global_block->AllOps()) {
    if (op->Type() == kFeedOpType) {
      std::string feed_target_name = op->Output("Out")[0];
      int idx = boost::get<int>(op->GetAttr("col"));
      SetFeedVariable(scope, *feed_targets[feed_target_name], feed_holder_name,
                      idx);
    }
  }

  if (!has_fetch_operators(global_block, fetch_targets, fetch_holder_name)) {
    // create fetch_holder variable
    auto* fetch_holder = global_block->Var(fetch_holder_name);
    fetch_holder->SetType(proto::VarType::FETCH_LIST);
    fetch_holder->SetPersistable(true);

    int i = 0;
    for (auto& fetch_target : fetch_targets) {
      std::string var_name = fetch_target.first;
      VLOG(3) << "fetch target's name: " << var_name;

      // append fetch op
      auto* op = global_block->AppendOp();
      op->SetType(kFetchOpType);
      op->SetInput("X", {var_name});
      op->SetOutput("Out", {fetch_holder_name});
      op->SetAttr("col", {static_cast<int>(i)});
      op->CheckAttrs();

      i++;
    }
  }

  Run(*copy_program, scope, 0, true, true);

  // obtain the data of fetch_targets from fetch_holder
  for (auto* op : global_block->AllOps()) {
    if (op->Type() == kFetchOpType) {
      std::string fetch_target_name = op->Input("X")[0];
      int idx = boost::get<int>(op->GetAttr("col"));
      *fetch_targets[fetch_target_name] =
          GetFetchVariable(*scope, fetch_holder_name, idx);
    }
  }

  delete copy_program;
}

ExecutorPrepareContext* Executor::Prepare(const ProgramDesc& program,
                                          int block_id) {
  auto* ctx = new ExecutorPrepareContext(program, block_id);
  PADDLE_ENFORCE_LT(static_cast<size_t>(block_id), program.Size());
  auto& block = program.Block(block_id);
  for (auto& op_desc : block.AllOps()) {
    ctx->ops_.push_back(OpRegistry::CreateOp(*op_desc));
  }
  return ctx;
}

void Executor::RunPreparedContext(ExecutorPrepareContext* ctx, Scope* scope,
                                  bool create_local_scope, bool create_vars) {
  auto& block = ctx->prog_.Block(ctx->block_id_);
  int block_id = block.ID();

  Scope* local_scope = scope;
  if (create_vars) {
    if (create_local_scope) {
      local_scope = &scope->NewScope();
      for (auto& var : block.AllVars()) {
        if (var->Name() == framework::kEmptyVarName) {
          continue;
        }

        if (var->Persistable()) {
          auto* ptr = scope->Var(var->Name());
          CreateTensor(ptr, var->GetType());
          VLOG(3) << "Create Variable " << var->Name()
                  << " global, which pointer is " << ptr;
        } else {
          auto* ptr = local_scope->Var(var->Name());
          CreateTensor(ptr, var->GetType());
          VLOG(3) << "Create Variable " << var->Name()
                  << " locally, which pointer is " << ptr;
        }
      }
    } else {
      for (auto& var : block.AllVars()) {
        auto* ptr = local_scope->Var(var->Name());
        CreateTensor(ptr, var->GetType());
        VLOG(3) << "Create variable " << var->Name() << ", which pointer is "
                << ptr;
      }
    }  // if (create_local_scope)
  }    // if (create_vars)

  platform::DeviceContextPool& pool = platform::DeviceContextPool::Instance();
  auto dev_ctx = pool.Get(place_);
  int dev_id_orig = boost::get<platform::CUDAPlace>(place_).GetDeviceId();
  int dev_id = dev_id_orig * 1000 + block_id;
  if (dev_id_orig == 0) {
    std::lock_guard<std::mutex> l2(reduce_mu);
    runned.clear();
  }

  std::mutex to_runs_mu;
  std::deque<OpDesc*> to_runs;
  std::unordered_map<int64_t, OpDesc*> running;
  int64_t cur_id = 0;
  for (auto& op_desc : block.AllOps()) {
    op_desc->Reset(dev_id);
    if (op_desc->IsReady(dev_id)) {
      to_runs.push_back(op_desc);
    }
  }

  std::unordered_map<std::string, OpDesc*> reduces;
  int cur_reduce = 0;

  while (true) {
    int64_t old_id = cur_id;
    cur_id++;
    OpDesc* op_desc = nullptr;
    bool is_all_running = false;
    bool is_too_many_running = false;
    {
      std::lock_guard<std::mutex> l(to_runs_mu);
      if (to_runs.empty()) {
        if (running.empty()) {
          break;
        } else {
          is_all_running = true;
        }
      } else {
        if (running.size() > 50) {
          is_all_running = true;
        } else {
          op_desc = to_runs.front();
          running[old_id] = op_desc;
          to_runs.pop_front();
        }
      }
    }
    if (is_all_running || is_too_many_running) {
      std::this_thread::sleep_for(std::chrono::microseconds(4));
      continue;
    }

    if (op_desc->UniqueName().find("ncclAllReduce") !=
        op_desc->UniqueName().npos) {
      if (dev_id_orig == 0) {
        auto op = paddle::framework::OpRegistry::CreateOp(*op_desc);
        // fprintf(stderr, "%s seq_start1 at %d at idx: %lu\n",
        //         op_desc->UniqueName().c_str(), dev_id_orig, runned.size());
        {
          std::lock_guard<std::mutex> l2(reduce_mu);
          runned.push_back(op_desc->UniqueName());
        }
        op->Run(*local_scope, place_);
        // fprintf(stderr, "%s seq_done1\n",
        //         op_desc->UniqueName().c_str());
        std::vector<OpDesc*> nexts = op_desc->GetRunnables(dev_id);
        {
          std::lock_guard<std::mutex> l(to_runs_mu);
          for (int i = 0; i < nexts.size(); ++i) {
            to_runs.push_back(nexts[i]);
          }
          running.erase(old_id);
        }
      } else {
        reduces[op_desc->UniqueName()] = op_desc;
        bool can_run = false;
        {
          std::lock_guard<std::mutex> l2(reduce_mu);
          can_run = cur_reduce < runned.size() &&
                    runned[cur_reduce] == op_desc->UniqueName() &&
                    reduces.find(runned[cur_reduce]) != reduces.end();
        }
        if (can_run) {
          // fprintf(stderr, "to run at idx: %d\n", cur_reduce);
          auto op = paddle::framework::OpRegistry::CreateOp(*op_desc);
          // fprintf(stderr, "%s seq_start2 at %d\n",
          //         op_desc->UniqueName().c_str(), dev_id_orig);
          op->Run(*local_scope, place_);
          // fprintf(stderr, "%s seq_done2\n", op_desc->UniqueName().c_str());
          std::vector<OpDesc*> nexts = op_desc->GetRunnables(dev_id);
          {
            std::lock_guard<std::mutex> l(to_runs_mu);
            for (int i = 0; i < nexts.size(); ++i) {
              to_runs.push_back(nexts[i]);
            }
            running.erase(old_id);
          }
          std::lock_guard<std::mutex> l2(reduce_mu);
          cur_reduce++;
        } else {
          std::lock_guard<std::mutex> l(to_runs_mu);
          running.erase(old_id);
          to_runs.push_back(op_desc);
        }
      }
      continue;
    }
    std::thread(
        [this, &to_runs, &to_runs_mu, op_desc, local_scope, dev_ctx, old_id,
            &running, dev_id] {
            OpDesc* desc = op_desc;
            platform::RecordEvent record_event(
                desc->UniqueName(), dev_ctx);
            auto op = paddle::framework::OpRegistry::CreateOp(*desc);
            // fprintf(stderr, "%s start3 at %d\n",
            //         desc->UniqueName().c_str(), dev_id);
            op->Run(*local_scope, place_);
            // fprintf(stderr, "%s done3\n", desc->UniqueName().c_str());

            std::vector<OpDesc*> nexts = desc->GetRunnables(dev_id);
            std::lock_guard<std::mutex> l(to_runs_mu);
            for (int i = 0; i < nexts.size(); ++i) {
              to_runs.push_back(nexts[i]);
            }
            running.erase(old_id);
        }).detach();
  }
  /*
  for (auto& op_desc : block.AllOps()) {
    auto op = paddle::framework::OpRegistry::CreateOp(*op_desc);
    VLOG(3) << place_ << " " << op->DebugStringEx(local_scope);
    op->Run(*local_scope, place_);

    std::thread t([&](){
        auto op = paddle::framework::OpRegistry::CreateOp(*op_desc);
        VLOG(3) << place_ << " " << op->DebugStringEx(local_scope);
        op->Run(*local_scope, place_);
    });
    t.join();
  }*/

  int64_t no_scheduled = 0;
  for (auto& op_desc : block.AllOps()) {
    if (!op_desc->Scheduled(dev_id)) {
      ++no_scheduled;
      fprintf(stderr, "%s not scheduled at %d\n",
              op_desc->UniqueName().c_str(), dev_id);
    }
  }

  if (create_vars && create_local_scope) {
    scope->DeleteScope(local_scope);
  }
  if (FLAGS_benchmark) {
    VLOG(2) << "-------------------------------------------------------";
    VLOG(2) << "Memory used after deleting local scope: "
            << memory::memory_usage(place_);
    VLOG(2) << "-------------------------------------------------------";
  }
}

}  // namespace framework
}  // namespace paddle

File: python/paddle/fluid/executor.py
#   Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import sys
import numpy as np
import contextlib
from framework import Program, default_main_program, Variable
from . import core

__all__ = [
    'Executor', 'global_scope', 'scope_guard', 'switch_scope', 'fetch_var'
]

g_scope = core.Scope()


def global_scope():
    return g_scope


def switch_scope(scope):
    global g_scope
    ex = g_scope
    g_scope = scope
    return ex


@contextlib.contextmanager
def scope_guard(scope):
    ex = switch_scope(scope)
    yield
    switch_scope(ex)


def as_numpy(tensor):
    if isinstance(tensor, list):
        return [as_numpy(t) for t in tensor]
    assert isinstance(tensor, core.LoDTensor)
    lod = tensor.lod()
    if len(lod) > 0:
        raise RuntimeError(
            "Some of your featched tensors hold LoD information. \
            They can not be completely cast to Python ndarray. \
            Please set the parameter 'return_numpy' as 'False' to \
            return LoDTensor itself directly.")
    return np.array(tensor)


def has_feed_operators(block, feed_targets, feed_holder_name):
    """ Check whether the block already has feed operators.

    Return false if the block does not have any feed operators.
    If some feed operators have been prepended to the block, check that
    the info contained in these feed operators matches the feed_targets
    and feed_holder_name. Raise exception when any mismatch is found.
    Return true when the block has feed operators with matching info.

    Args:
        block: a block instance (typically global block of a program)
        feed_targets: a dictionary of {feed_target_name: feed_target_data}
        feed_holder_name: the name of the variable that holds the data of
            all feed targets. The type of this feed_holder variable is
            FEED_MINIBATCH, which is essentially vector<LoDTensor>.

    Returns:
        A boolean value that indicates whether a block has feed operators
        that match the info contained in feed_targets and feed_holder_name.
    """

    feed_count = 0
    for op in block.ops:
        if op.desc.type() == 'feed':
            feed_count += 1
            assert op.desc.input('X')[0] == feed_holder_name
            feed_target_name = op.desc.output('Out')[0]
            if feed_target_name not in feed_targets:
                raise Exception("'feed_targets' does not have {} variable".
                                format(feed_target_name))
        else:
            break
    if feed_count > 0 and feed_count != len(feed_targets):
        raise Exception(
            "Feed operators in program desc do not match 'feed_targets'")
    return feed_count > 0


def has_fetch_operators(block, fetch_targets, fetch_holder_name):
    """ Check whether the block already has fetch operators.

    Return false if the block does not have any fetch operators.
    If some fetch operators have been appended to the block, check that
    the info contained in these fetch operators matches the fetch_targets
    and fetch_holder_name. Raise exception when any mismatch is found.
    Return true when the block has fetch operators with matching info.

    Args:
        block: a block instance (typically global block of a program)
        fetch_targets: a dictionary of {fetch_target_name: fetch_target_data}
        fetch_holder_name: the name of the variable that holds the data of
            all fetch targets. The type of this fetch_holder variable is
            FETCH_LIST, which is essentially vector<LoDTensor>.

    Return:
        A boolean value that indicates whether a block has fetch operators
        that match the info contained in fetch_targets and fetch_holder_name.
    """

    fetch_count = 0
    for op in block.ops:
        if op.desc.type() == 'fetch':
            fetch_count += 1
            assert op.desc.output('Out')[0] == fetch_holder_name
            fetch_target_name = op.desc.input('X')[0]
            if fetch_target_name not in [
                    var.desc.name() for var in fetch_targets
            ]:
                raise Exception("'fetch_targets' does not have {} variable".
                                format(fetch_target_name))
            idx = op.desc.attr('col')
            assert fetch_target_name == fetch_targets[idx].desc.name()
    if fetch_count > 0 and fetch_count != len(fetch_targets):
        raise Exception(
            "Fetch operators in program desc do not match 'fetch_targets'")
    return fetch_count > 0


def fetch_var(name, scope=None, return_numpy=True):
    """
    Fetch the value of the variable with the given name from the given scope
    Args:
        name(str): name of the variable. Typically, only persistable variables
            can be found in the scope used for running the program.
        scope(core.Scope|None): scope object. It should be the scope where
            you pass to Executor.run() when running your program.
            If None, global_scope() will be used.
        return_numpy(bool): whether convert the tensor to numpy.ndarray
    Returns:
       LodTensor|numpy.ndarray
    """
    assert isinstance(name, str)
    if scope is None:
        scope = global_scope()
    assert isinstance(scope, core.Scope)

    var = global_scope().find_var(name)
    assert var is not None, (
        "Cannot find " + name + " in scope. Perhaps you need to make the"
        " variable persistable by using var.persistable = True in your"
        " program.")
    tensor = var.get_tensor()
    if return_numpy:
        tensor = as_numpy(tensor)
    return tensor


def get_program_cache_key(feed, fetch_list):
    feed_var_names = feed.keys()

    def to_name_str(var):
        if isinstance(var, Variable):
            return var.desc.name()
        elif isinstance(var, str):
            return var
        else:
            raise TypeError(str(var) + " should be Variable or str")

    fetch_var_names = map(to_name_str, fetch_list)

    return str(feed_var_names + fetch_var_names)



def find_loops():
    pass

def analyze_dependency(program):
    output_to_op = dict()
    op_idx = dict()
    ncclInit = None

    idx = 0
    should_print0 = False
    for block in program.blocks:
        for op in block.ops:
            # if 'ncclInit' in op.desc.unique_name():
            #     should_print0 = True
            op.desc.set_unique_name(block.idx)
            op_idx[op.desc.unique_name()] = idx
            idx += 1
            if 'ncclInit' in op.desc.unique_name():
                assert not ncclInit
                ncclInit = op

            for output_name in op.output_names:
                for output_arg in op.output(output_name):
                    if should_print0:
                        sys.stderr.write('ncclInit output %s\n' % output_arg)
                    # if output_arg in output_to_op:
                    #     sys.stderr.write('!!!!!%s duplicated\n' % output_arg)
                    if output_arg == 'fetch':
                        sys.stderr.write('!!!!!%s output from %s\n' %
                                         (output_arg, op.type))
                    if output_arg not in output_to_op:
                        output_to_op[output_arg] = []
                    output_to_op[output_arg].append(op)
            should_print0 = False

    should_print = False
    for block in program.blocks:
        for op in block.ops:
            # if "nccl" in op.desc.unique_name():
            #     sys.stderr.write('nccl: %s\n' % op.desc.unique_name())
            #     should_print = True
            if ncclInit and 'parallel_do_grad' in op.desc.unique_name():
                ncclInit.desc.add_dependent(op.desc)
                sys.stderr.write('%s add dependent %s\n' %
                                 (ncclInit.desc.unique_name(),
                                  op.desc.unique_name()))
            for input_name in op.input_names:
                for input_arg in op.input(input_name):
                    if should_print:
                        sys.stderr.write('nccl input_arg: %s\n' % input_arg)
                    if input_arg not in output_to_op:
                        if should_print:
                            sys.stderr.write('1111\n')
                        # sys.stderr.write(
                        #     '!!!!!%s not output op\n' % input_arg)
                        continue
                    for output_op in output_to_op[input_arg]:
                        if (op_idx[output_op.desc.unique_name()] >=
                            op_idx[op.desc.unique_name()]):
                            if should_print:
                                sys.stderr.write(
                                    '22222 %s %s %d %d\n' % (
                                        op.desc.unique_name(),
                                        output_op.desc.unique_name(),
                                        op_idx[output_op.desc.unique_name()],
                                        op_idx[op.desc.unique_name()]))
                            continue
                        if output_op.block.idx != op.block.idx:
                            if should_print:
                                sys.stderr.write('33333 %d %d\n' %
                                                 (output_op.block.idx,
                                                  op.block.idx))
                            continue
                        # sys.stderr.write('adding %s depends on %s\n' % (
                        #     op.desc.unique_name(),
                        #     output_to_op[input_arg].desc.unique_name()))
                        output_op.desc.add_dependent(op.desc)
            should_print = False
    """        
    op_nexts = dict()
    op_previous = dict()
    for block in program.blocks:
        for op in block.ops:
            nexts = op.desc.all_dep_ops()
            op_nexts[op.desc.unique_name()] = nexts
            for n in nexts:
                if n not in op_previous:
                    op_previous[n] = []
                op_previous[n].append(op.desc.unique_name())
    for op, previous in op_previous.iteritems():
        sys.stderr.write('%s previous %s\n' % (op, ','.join(previous)))
    """
    # sys.stderr.write('%s' % program)


class Executor(object):
    def __init__(self, places):
        if not isinstance(places, list) and not isinstance(places, tuple):
            places = [places]

        act_places = []
        for each in places:
            p = core.Place()
            p.set_place(each)
            act_places.append(p)

        # TODO(dzhwinter) : only use the first place
        self.executor = core.Executor(act_places[0])
        self.places = places
        self.program_caches = dict()
        self._analyzed = set()

    def aslodtensor(self, data):
        def accumulate(data):
            if not isinstance(data, list):
                return 1
            return sum([accumulate(sub) for sub in data])

        def parselod(data):
            seq_lens = [accumulate(seq) for seq in data]
            cur_len = 0
            lod = [cur_len]
            for l in seq_lens:
                cur_len += l
                lod.append(cur_len)
            return lod

        assert len(self.places) != 0
        if not isinstance(data, list):
            # pure tensor case
            tensor = core.LoDTensor()
            tensor.set(data, self.places[0])
            return tensor
        else:
            raise RuntimeError("Current implementation lacks unittests")
            # lodtensor case
            lod = []
            if not isinstance(data[0], list):
                lod.append(parselod(data))
                flattened_data = np.concatenate(data, axis=0).astype("int64")
            else:
                while isinstance(data[0], list):
                    lod.append(parselod(seq))
                    flattened_data = [item for seq in data for item in seq]
                    data = flattened_data
                flattened_data = np.concatenate(data, axis=0).astype("int64")
            flattened_data = flattened_data.reshape([len(flattened_data), 1])
            tensor = core.LoDTensor()
            tensor.set(flattened_data, self.places[0])
            tensor.set_lod(lod)
            return tensor

    def run(self,
            program=None,
            feed=None,
            fetch_list=None,
            feed_var_name='feed',
            fetch_var_name='fetch',
            scope=None,
            return_numpy=True,
            use_program_cache=False):
        """ Run program by this Executor. Feed data by feed map, fetch result by fetch_list.

        Python executor takes a program, add feed operators and fetch operators to this program according
        to feed map and fetch_list. Feed map provides input data for the program. fetch_list provides
        the variables(or names) that user want to get after program run. Note: the executor will run all
        operators in the program but not only the operators dependent by the fetch_list

        :param program: the program that need to run, if not provied, then default_main_program will be used.
        :param feed: feed variable map, e.g. {"image": ImageData, "label": LableData}
        :param fetch_list: a list of variable or variable names that user want to get, run will return them according
        to this list.
        :param feed_var_name: the name for the input variable of feed Operator.
        :param fetch_var_name: the name for the output variable of feed Operator.
        :param scope: the scope used to run this program, you can switch it to different scope. default is global_scope
        :param return_numpy: if convert the fetched tensor to numpy
        :param use_program_cache: set use_program_cache to true if program not changed compare to the last step.
        :return: result according to fetch_list.
        """
        if feed is None:
            feed = {}
        if not isinstance(feed, dict):
            raise TypeError("feed should be a map")
        if fetch_list is None:
            fetch_list = []

        if program is None:
            program = default_main_program()

        if not isinstance(program, Program):
            raise TypeError()

        if scope is None:
            scope = global_scope()

        program_cache = None
        program_cache_key = get_program_cache_key(feed, fetch_list)

        if use_program_cache:
            # find program cache by cache_key
            program_cache = self.program_caches.get(program_cache_key, None)
            # TODO(qiao): Should check program_cache and program are exactly the same.
        else:
            self.program_caches.pop(program_cache_key, None)

        if program_cache is None:
            program_cache = program.clone()


            if use_program_cache:
                self.program_caches[program_cache_key] = program_cache

            global_block = program_cache.global_block()

            if feed_var_name in global_block.vars:
                feed_var = global_block.var(feed_var_name)
            else:
                feed_var = global_block.create_var(
                    name=feed_var_name,
                    type=core.VarDesc.VarType.FEED_MINIBATCH,
                    persistable=True)

            if fetch_var_name in global_block.vars:
                fetch_var = global_block.var(fetch_var_name)
            else:
                fetch_var = global_block.create_var(
                    name=fetch_var_name,
                    type=core.VarDesc.VarType.FETCH_LIST,
                    persistable=True)

            # prepend feed operators
            if not has_feed_operators(global_block, feed, feed_var_name):
                for i, name in enumerate(feed):
                    out = global_block.var(name)
                    global_block.prepend_op(
                        type='feed',
                        inputs={'X': [feed_var]},
                        outputs={'Out': [out]},
                        attrs={'col': i})

            # append fetch_operators
            if not has_fetch_operators(global_block, fetch_list,
                                       fetch_var_name):
                for i, var in enumerate(fetch_list):
                    assert isinstance(var, Variable) or isinstance(var, str), (
                        "Wrong type for fetch_list[%s]: %s" % (i, type(var)))
                    global_block.append_op(
                        type='fetch',
                        inputs={'X': [var]},
                        outputs={'Out': [fetch_var]},
                        attrs={'col': i})

        # feed var to framework
        for op in program_cache.global_block().ops:
            if op.desc.type() == 'feed':
                feed_target_name = op.desc.output('Out')[0]
                cur_feed = feed[feed_target_name]
                if not isinstance(cur_feed, core.LoDTensor):
                    cur_feed = self.aslodtensor(cur_feed)
                idx = op.desc.attr('col')
                core.set_feed_variable(scope, cur_feed, feed_var_name, idx)
            else:
                break

        if not program_cache in self._analyzed:
            analyze_dependency(program_cache)
            self._analyzed.add(program_cache)
            sys.stderr.write('analyzed a program\n')

        self.executor.run(program_cache.desc, scope, 0, True, True)
        outs = [
            core.get_fetch_variable(scope, fetch_var_name, i)
            for i in xrange(len(fetch_list))
        ]
        if return_numpy:
            outs = as_numpy(outs)
        return outs
